text/0059-otlp-trace-data-format.md

```diff
@@ -0,0 +1,348 @@
+    // as opposed to an operations happening at the boundaries. Default value.
+    INTERNAL = 1;
+
+    // Indicates that the span covers server-side handling of an RPC or other
+    // remote network request.
+    SERVER = 2;
+
+    // Indicates that the span describes a request to some remote service.
+    CLIENT = 3;
+
+    // Indicates that the span describes a producer sending a message to a broker.
+    // Unlike CLIENT and SERVER, there is often no direct critical path latency relationship
+    // between producer and consumer spans. A PRODUCER span ends when the message was accepted
+    // by the broker while the logical processing of the message might span a much longer time.
+    PRODUCER = 4;
+
+    // Indicates that the span describes consumer receiving a message from a broker.
+    // Like the PRODUCER kind, there is often no direct critical path latency relationship
+    // between producer and consumer spans.
+    CONSUMER = 5;
+  }
+
+  // kind field distinguishes between spans generated in a particular context. For example,
+  // two spans with the same name may be distinguished using `CLIENT` (caller)
+  // and `SERVER` (callee) to identify network latency associated with the span.
+  SpanKind kind = 7;
+
+  // start_time_unixnano is the start time of the span. On the client side, this is the time
+  // kept by the local machine where the span execution starts. On the server side, this
+  // is the time when the server's application handler starts running.
+  //
+  // This field is semantically required and it is expected that end_time >= start_time.
+  //
+  // This field is required.
+  int64 start_time_unixnano = 8;
+
+  // end_time_unixnano is the end time of the span. On the client side, this is the time
+  // kept by the local machine where the span execution ends. On the server side, this
+  // is the time when the server application handler stops running.
+  //
+  // This field is semantically required and it is expected that end_time >= start_time.
+  //
+  // This field is required.
+  int64 end_time_unixnano = 9;
+
+  // attributes is a collection of attribute key/value pairs. The value can be a string,
+  // an integer, a double or the Boolean values `true` or `false`. Note, global attributes
+  // like server name can be set as tags using resource API. Examples of attributes:
+  //
+  //     "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
+  //     "/http/server_latency": 300
+  //     "abc.com/myattribute": true
+  //     "abc.com/score": 10.239
+  repeated AttributeKeyValue attributes = 10;
+
+  // dropped_attributes_count is the number of attributes that were discarded. Attributes
+  // can be discarded because their keys are too long or because there are too many
+  // attributes. If this value is 0, then no attributes were dropped.
+  int32 dropped_attributes_count = 11;
```

Comment:
feels like this should be DRYed into a `struct { attributes, dropped_count }` that can be reused in other entities like Resource, TimedEvent. The naming will be odd though, `span.attributes.attributes`

Reply:
I agree it would be more DRY, but unfortunately it has measurable performance impact.

Reply:
Should we use fixed64?

Reply:
I tried fixed32 and don't see any measurable difference in speed (only slight increase of uncompressed wire size).

I can change to fixed if there is any argument in favour of it.